### PR TITLE
virtio-blk: surface krun_add_disk2 block_id as device ID/serial

### PR DESCRIPTION
**Pull request by dunnry**
VIRTIO_BLK_T_GET_ID previously returned a file-derived image id, so guests saw arbitrary serials under /sys/block/*/serial and could not locate disks by the block_id passed to krun_add_disk2. Use the supplied block id as the device ID so exact-serial discovery works.

This path also exists on stable-1.19.x and stable-1.18.x, so this may be suitable for the cherry-pick-1.19.5 and cherry-pick-1.18.2 labels.

Tests:
- cargo fmt --all -- --check
- git diff --check origin/main..HEAD
- cargo test -p krun-devices --features blk disk_image_id

### CODE REVIEW

File: src/devices/src/virtio/block/device.rs [hunk 1]
```diff
@@ -469,3 +479,47 @@ impl VirtioDevice for Block {
...
+
+        let image_id = DiskProperties::build_disk_image_id(backing.as_file(), long_id);
+
+        assert_eq!(image_id.len(), VIRTIO_BLK_ID_BYTES as usize);
+        assert_eq!(
+            &image_id,
+            &long_id.as_bytes()[..VIRTIO_BLK_ID_BYTES as usize]
+        );
+    }
+}
```
Comment (mz-pdm) [inline src/devices/src/virtio/block/device.rs:538]: Would it make sense to add a test for the empty block id case?  To check that the result has the right size and nothing bad (like an unhandled error) happens?

**Review by mz-pdm — DISMISSED** [review]
LGTM, just one optional comment.

File: src/devices/src/virtio/block/device.rs (hunk 1 above)
Comment (dunnry) [reply src/devices/src/virtio/block/device.rs:538]: Added an empty block_id test that exercises the fallback path and asserts the returned ID has the virtio-blk fixed size.

**Review by mz-pdm — APPROVED** [review]

**Review by slp — APPROVED** [review]
LGTM, thanks!

**Comment by mtjhrc** [comment]
@slp I marked this as "cherry-pick-1.19.5" as mentioned in the PR description so we don't forget about this. Not sure if that is correct assessment of whether this can break consumers, but we can still discuss that when compiling PRs for the release.

I think @dunnry would want this in a stable release.

**Comment by slp** [comment]
> @slp I marked this as "cherry-pick-1.19.5" as mentioned in the PR description so we don't forget about this. Not sure if that is correct assessment of whether this can break consumers, but we can still discuss that when compiling PRs for the release.

Thanks!
